# Import Library

In [1]:
import pandas as pd
import numpy as np

# Read Dataset

In [2]:
import os
import kagglehub

# Download latest version to define 'path'
path = kagglehub.dataset_download("kalkulasi/financial-statement-data-idx-2020-2023")

# List files in the downloaded directory
print(os.listdir(path))

# Construct the full path to the CSV file
csv_file_path = os.path.join(path, 'combined_financial_data_idx.csv')

# Read the CSV file into a pandas DataFrame
df = pd.read_csv(csv_file_path)

print("Path to dataset files:", path)
display(df.head())

100%|██████████| 1.74M/1.74M [00:00<00:00, 89.0MB/s]

Extracting files...
['combined_financial_data_idx.csv']


Path to dataset files: /root/.cache/kagglehub/datasets/kalkulasi/financial-statement-data-idx-2020-2023/versions/1


symbol                                     account type          2020  \
0   AALI                            Accounts Payable   BS  7.702640e+11   
1   AALI                         Accounts Receivable   BS  7.658490e+11   
2   AALI                    Accumulated Depreciation   BS -1.092095e+13   
3   AALI                  Additional Paid In Capital   BS  3.878995e+12   
4   AALI  Allowance For Doubtful Accounts Receivable   BS -2.426100e+10   

           2021          2022          2023  
0  1.026717e+12  1.224423e+12  8.420640e+11  
1  4.581350e+11  8.487700e+11  6.744870e+11  
2 -1.213381e+13 -1.330375e+13 -1.443685e+13  
3  3.878995e+12  3.878995e+12  3.878995e+12  
4 -2.454300e+10 -2.705700e+10 -2.651600e+10

In [3]:
pd.options.display.float_format = '{:.2f}'.format
display(df.head())

symbol                                     account type               2020  \
0   AALI                            Accounts Payable   BS    770264000000.00   
1   AALI                         Accounts Receivable   BS    765849000000.00   
2   AALI                    Accumulated Depreciation   BS -10920948000000.00   
3   AALI                  Additional Paid In Capital   BS   3878995000000.00   
4   AALI  Allowance For Doubtful Accounts Receivable   BS    -24261000000.00   

                2021               2022               2023  
0   1026717000000.00   1224423000000.00    842064000000.00  
1    458135000000.00    848770000000.00    674487000000.00  
2 -12133813000000.00 -13303749000000.00 -14436847000000.00  
3   3878995000000.00   3878995000000.00   3878995000000.00  
4    -24543000000.00    -27057000000.00    -26516000000.00

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89243 entries, 0 to 89242
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   symbol   89243 non-null  object 
 1   account  89243 non-null  object 
 2   type     89243 non-null  object 
 3   2020     83988 non-null  float64
 4   2021     84474 non-null  float64
 5   2022     83634 non-null  float64
 6   2023     69831 non-null  float64
dtypes: float64(4), object(3)
memory usage: 4.8+ MB


In [ ]:
df.sample(10)

symbol                                            account type  \
49658   LPPF                             Receiptsfrom Customers   CF   
4648    ANTM                   Net Income Continuous Operations   IS   
73078   SIPD                         Non Current Prepaid Assets   BS   
38019   IBST                                      Tax Provision   IS   
67583   PTPP                          Capital Lease Obligations   BS   
83686   TRIM  Net Income From Continuing Operation Net Minor...   IS   
39614   IMPC                            Classesof Cash Payments   CF   
67178   PTBA               Goodwill And Other Intangible Assets   BS   
56870   MPPA                                  Repayment Of Debt   CF   
81562   TKIM                                     Prepaid Assets   BS   

                   2020              2021              2022              2023  
49658  9505393000000.00 11333432000000.00 13664097000000.00 13865427000000.00  
4648   1149354000000.00  1861740000000.00  3820964000000.00  3077648000000.00  
73078     1346000000.00     1099000000.00     1122000000.00     1054000000.00  
38019    24990246564.00    12324553179.00    23255173376.00               NaN  
67583   223309222269.00   335829717726.00   389007241039.00   356128305728.00  
83686    28262922000.00    51892895000.00   177500956000.00   162462603000.00  
39614 -1413476211224.00 -1870155952107.00 -2349916087933.00 -2083019732181.00  
67178   102077000000.00   102077000000.00   102077000000.00   102077000000.00  
56870               NaN  -160000000000.00  -120000000000.00   -80000000000.00  
81562       52171000.00       56360000.00       32874000.00       38529000.00

# Data Cleaning

In [ ]:
df.duplicated().sum()

np.int64(0)

In [ ]:
df.isna().sum()

symbol         0
account        0
type           0
2020        5255
2021        4769
2022        5609
2023       19412
dtype: int64

In [ ]:
# untuk melihat value kolom
for column in df.columns:
    print(f"============= {column} =================")
    display(df[column].value_counts())
    print()


============= symbol =================


symbol
TLKM    199
JPFA    189
INDY    188
LPKR    184
MAPI    184
       ... 
BACA     97
BINA     96
LCGP     96
PNBS     94
BANK     92
Name: count, Length: 604, dtype: int64


============= account =================


account
Total Tax Payable                          604
Total Revenue                              604
Total Liabilities Net Minority Interest    604
Total Equity Gross Minority Interest       604
Total Assets                               604
                                          ... 
Other Equity Adjustments                     1
Other Gand A                                 1
Payables And Accrued Expenses                1
Provision For Doubtful Accounts              1
Salaries And Wages                           1
Name: count, Length: 264, dtype: int64


============= type =================


type
BS    38675
IS    29487
CF    21081
Name: count, dtype: int64


============= 2020 =================


2020
0.00             3781
0.25              198
1.00               40
-1.00              35
5000000000.00      29
                 ... 
17132477.00         1
73940.00            1
88169.00            1
1353415.00          1
2488901.00          1
Name: count, Length: 59804, dtype: int64


============= 2021 =================


2021
0.00                 3876
0.22                  151
-1.00                  31
1000000000.00          31
5000000000.00          24
                     ... 
233295000000.00         1
9840000000.00           1
439461000000.00         1
-33379000000.00         1
-4809167000000.00       1
Name: count, Length: 60308, dtype: int64


============= 2022 =================


2022
0.00                 3550
0.22                  129
-1.00                  29
1.00                   21
100000000000.00        17
                     ... 
1188123000000.00        1
7177503000000.00        1
3918100000000.00        1
-3957798000000.00       1
4538398000000.00        1
Name: count, Length: 60103, dtype: int64


============= 2023 =================


2023
0.00                  2453
0.22                   123
1.00                    20
1000000000.00           18
-1.00                   18
                      ... 
4409263000000.00         1
-95712000000.00          1
-99883323000000.00       1
2100968000000.00         1
53700000000.00           1
Name: count, Length: 51369, dtype: int64

In [ ]:
# 1Interpolasi antar tahun (untuk kolom numerik)
df[['2020', '2021', '2022', '2023']] = df[['2020', '2021', '2022', '2023']].interpolate(
    method='linear', axis=1)

# 2Forward fill / backward fill untuk nilai di awal / akhir
df[['2020', '2021', '2022', '2023']] = df[['2020', '2021', '2022', '2023']].fillna(method='ffill', axis=1)
df[['2020', '2021', '2022', '2023']] = df[['2020', '2021', '2022', '2023']].fillna(method='bfill', axis=1)

# 3Drop baris yang tetap NaN di semua tahun
df = df.dropna(subset=['2020', '2021', '2022', '2023'], how='all')

df.head()

/tmp/ipython-input-3254709397.py:6: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df[['2020', '2021', '2022', '2023']] = df[['2020', '2021', '2022', '2023']].fillna(method='ffill', axis=1)
/tmp/ipython-input-3254709397.py:7: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df[['2020', '2021', '2022', '2023']] = df[['2020', '2021', '2022', '2023']].fillna(method='bfill', axis=1)


symbol                                     account type               2020  \
0   AALI                            Accounts Payable   BS    770264000000.00   
1   AALI                         Accounts Receivable   BS    765849000000.00   
2   AALI                    Accumulated Depreciation   BS -10920948000000.00   
3   AALI                  Additional Paid In Capital   BS   3878995000000.00   
4   AALI  Allowance For Doubtful Accounts Receivable   BS    -24261000000.00   

                2021               2022               2023  
0   1026717000000.00   1224423000000.00    842064000000.00  
1    458135000000.00    848770000000.00    674487000000.00  
2 -12133813000000.00 -13303749000000.00 -14436847000000.00  
3   3878995000000.00   3878995000000.00   3878995000000.00  
4    -24543000000.00    -27057000000.00    -26516000000.00

- Interpolasi Linier (interpolate(method='linear', axis=1)): Mengisi nilai NaN di antara tahun-tahun yang memiliki data. Misalnya, jika data tahun 2020 dan 2022 ada tetapi 2021 hilang, interpolasi linier akan mengisi 2021 dengan nilai di tengah antara 2020 dan 2022. Ini berasumsi ada tren linier antara titik data yang diketahui.

- Forward Fill (fillna(method='ffill', axis=1)): Mengisi nilai NaN dengan nilai terakhir yang valid di baris yang sama. Berguna jika Anda berasumsi nilai keuangan tetap sama dari tahun sebelumnya jika tidak dilaporkan.

- Backward Fill (fillna(method='bfill', axis=1)): Mengisi nilai NaN dengan nilai valid pertama setelahnya di baris yang sama. Berguna untuk asumsi nilai keuangan akan sama dengan nilai yang dilaporkan di tahun berikutnya.

- Drop Rows (dropna(subset=..., how='all')): Setelah mencoba mengisi nilai, kode ini menghapus baris di mana semua nilai untuk tahun 2020 hingga 2023 masih NaN. 

In [ ]:
df.isna().sum()

symbol     0
account    0
type       0
2020       0
2021       0
2022       0
2023       0
dtype: int64

In [ ]:
df.describe()

2020                2021                2022  \
count            89243.00            89243.00            89243.00   
mean     1130056558779.59    1331810647475.24    1417131925626.94   
std     14953088039086.50   16715334461876.22   18307408717635.50   
min   -412793414000000.00 -482635842000000.00 -561001621000000.00   
25%                  0.00                0.00                0.00   
50%         4647000000.00       5678000000.00       6382787788.00   
75%       191481449009.00     232316967420.00     258688000000.00   
max   1610065344000000.00 1725611128000000.00 1992544687000000.00   

                     2023  
count            89243.00  
mean     1441638666180.87  
std     19269785937717.25  
min   -570917643000000.00  
25%                  0.00  
50%         6520204175.00  
75%       257559897500.00  
max   2174219449000000.00

- Count (Jumlah Data): Semua kolom tahun sekarang memiliki jumlah data yang sama (89243), yang mengkonfirmasi bahwa nilai-nilai NaN telah berhasil ditangani dan tidak ada baris yang dihapus secara massal kecuali yang memang kosong di semua tahun.
Mean (Rata-rata): Nilai rata-rata menunjukkan tren kenaikan dari tahun 2020 hingga 2023 (dari sekitar 1.13 triliun menjadi 1.44 triliun). Ini mengindikasikan bahwa, secara rata-rata di seluruh akun keuangan dan perusahaan, nilai-nilai cenderung meningkat selama periode ini.
- Std (Standar Deviasi): Standar deviasi sangat besar dan meningkat dari tahun ke tahun. Ini menunjukkan variasi yang sangat besar dalam nilai-nilai keuangan di seluruh data Anda. Ada perbedaan yang signifikan antara nilai terendah dan tertinggi, serta banyak nilai yang tersebar jauh dari rata-rata. Ini wajar dalam data keuangan yang mencakup perusahaan besar dan kecil dengan akun keuangan yang sangat bervariasi nilainya.
Min (Nilai Minimum): Ada nilai minimum negatif yang cukup besar, menunjukkan bahwa ada akun keuangan (seperti akumulasi penyusutan atau beberapa jenis liabilitas/ekuitas) yang memang bernilai negatif. Nilai minimum ini juga cenderung menurun (menjadi lebih negatif) dari tahun ke tahun.
- 25% (Kuartil Pertama): Seperti yang kita diskusikan, 25% data terendah di setiap tahun bernilai 0.00. Ini menegaskan bahwa seperempat dari data memiliki nilai nol.
- 50% (Median/Kuartil Kedua): Nilai median (atau kuartil kedua) jauh lebih rendah daripada rata-rata (sekitar 4.6 miliar hingga 6.5 miliar). Ini menunjukkan bahwa distribusi data sangat miring (skewed) ke kanan, dengan banyak nilai yang relatif kecil dan sejumlah kecil nilai yang sangat besar (outlier) yang menarik rata-rata ke atas. Median memberikan gambaran yang lebih baik tentang "nilai tipikal" dalam data dibandingkan rata-rata. Nilai median juga menunjukkan sedikit peningkatan dari tahun ke tahun.
- 75% (Kuartil Ketiga): Nilai kuartil ketiga (sekitar 191 miliar hingga 257 miliar) jauh di atas median tetapi masih jauh di bawah rata-rata maksimum. Ini menunjukkan bahwa sebagian besar data (dari 25% hingga 75%) berada dalam rentang nilai yang lebih terkonsentrasi dibandingkan keseluruhan rentang data.
- Max (Nilai Maksimum): Ada nilai maksimum yang sangat besar (mencapai triliunan), yang merupakan outlier positif yang signifikan dan berkontribusi besar terhadap nilai rata-rata yang tinggi dan standar deviasi yang besar.

### Data Transformation for Visualization (Long Format)

In [ ]:
# Melt the DataFrame to transform it from wide to long format
# This is ideal for visualizing time-series data in Streamlit
df_long = df.melt(id_vars=['symbol', 'account', 'type'],
                  var_name='Year',
                  value_name='Value')

# Convert 'Year' column to integer type for proper sorting and plotting
df_long['Year'] = df_long['Year'].astype(int)

print("DataFrame in long format:")
display(df_long.head())

DataFrame in long format:


symbol                                     account type  Year  \
0   AALI                            Accounts Payable   BS  2020   
1   AALI                         Accounts Receivable   BS  2020   
2   AALI                    Accumulated Depreciation   BS  2020   
3   AALI                  Additional Paid In Capital   BS  2020   
4   AALI  Allowance For Doubtful Accounts Receivable   BS  2020   

               Value  
0    770264000000.00  
1    765849000000.00  
2 -10920948000000.00  
3   3878995000000.00  
4    -24261000000.00

In [ ]:
df_long.sample(10)

symbol                                       account type  Year  \
128530   IMAS           Diluted NI Availto Com Stockholders   IS  2021   
78140    SULI                    Reconciled Cost Of Revenue   IS  2020   
240669   PBRX                                     Gross PPE   BS  2022   
263623   UNSP        Total Unusual Items Excluding Goodwill   IS  2022   
206602   DYAN                                  Common Stock   BS  2022   
332163   PNBS   Long Term Debt And Capital Lease Obligation   BS  2023   
49864    LSIP                     Gross Accounts Receivable   BS  2020   
148813   NCKL  Other Cash Paymentsfrom Operating Activities   CF  2021   
182078   AMAR                      Gain On Sale Of Security   IS  2022   
221974   JARR          Total Equity Gross Minority Interest   BS  2022   

                  Value  
128530 -264778000000.00  
78140       55597323.00  
240669     217093466.00  
263623 1797201000000.00  
206602  427296427900.00  
332163   13510362000.00  
49864   214240000000.00  
148813 -957466814094.00  
182078   13920000000.00  
221974 1274264114958.00

In [ ]:
print("Information about the long format DataFrame:")
df_long.info()

Information about the long format DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 356972 entries, 0 to 356971
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   symbol   356972 non-null  object 
 1   account  356972 non-null  object 
 2   type     356972 non-null  object 
 3   Year     356972 non-null  int64  
 4   Value    356972 non-null  float64
dtypes: float64(1), int64(1), object(3)
memory usage: 13.6+ MB


In [ ]:
print("Descriptive statistics for the long format DataFrame:")
df_long.describe()

Descriptive statistics for the long format DataFrame:


Year               Value
count 356972.00           356972.00
mean    2021.50    1330159449515.66
std        1.12   17389170361581.40
min     2020.00 -570917643000000.00
25%     2020.75                0.00
50%     2021.50       5745016334.50
75%     2022.25     234375000000.00
max     2023.00 2174219449000000.00

### Data Exploration: Unique Categories

In [ ]:
print("Unique values in 'symbol' column:")
print(df_long['symbol'].nunique())
display(df_long['symbol'].value_counts().head())

print("\nUnique values in 'account' column:")
print(df_long['account'].nunique())
display(df_long['account'].value_counts().head())

print("\nUnique values in 'type' column:")
print(df_long['type'].nunique())
display(df_long['type'].value_counts())

Unique values in 'symbol' column:
604


symbol
TLKM    796
JPFA    756
INDY    752
LPKR    736
MAPI    736
Name: count, dtype: int64


Unique values in 'account' column:
264


account
Total Tax Payable                          2416
Total Revenue                              2416
Total Liabilities Net Minority Interest    2416
Total Equity Gross Minority Interest       2416
Total Assets                               2416
Name: count, dtype: int64


Unique values in 'type' column:
3


type
BS    154700
IS    117948
CF     84324
Name: count, dtype: int64

Dari hasil di atas, kita melihat bahwa ada:
- **604 simbol perusahaan** yang berbeda, dengan `TLKM` dan `JPFA` memiliki entri terbanyak.
- **199 akun keuangan** yang berbeda, seperti `Common Stock`, `Retained Earnings`, dll.
- Hanya **3 tipe laporan** (`BS` untuk Balance Sheet, `IS` untuk Income Statement, `CF` untuk Cash Flow Statement).

In [ ]:
df.head()

symbol                                     account type               2020  \
0   AALI                            Accounts Payable   BS    770264000000.00   
1   AALI                         Accounts Receivable   BS    765849000000.00   
2   AALI                    Accumulated Depreciation   BS -10920948000000.00   
3   AALI                  Additional Paid In Capital   BS   3878995000000.00   
4   AALI  Allowance For Doubtful Accounts Receivable   BS    -24261000000.00   

                2021               2022               2023  
0   1026717000000.00   1224423000000.00    842064000000.00  
1    458135000000.00    848770000000.00    674487000000.00  
2 -12133813000000.00 -13303749000000.00 -14436847000000.00  
3   3878995000000.00   3878995000000.00   3878995000000.00  
4    -24543000000.00    -27057000000.00    -26516000000.00

In [ ]:
df_long.head()

symbol                                     account type  Year  \
0   AALI                            Accounts Payable   BS  2020   
1   AALI                         Accounts Receivable   BS  2020   
2   AALI                    Accumulated Depreciation   BS  2020   
3   AALI                  Additional Paid In Capital   BS  2020   
4   AALI  Allowance For Doubtful Accounts Receivable   BS  2020   

               Value  
0    770264000000.00  
1    765849000000.00  
2 -10920948000000.00  
3   3878995000000.00  
4    -24261000000.00

# Save Data

In [ ]:
df.to_csv('cleaned_financial_data_wide.csv', index=False, encoding='utf-8')
print("Data yang sudah dibersihkan (format wide) berhasil disimpan ke 'cleaned_financial_data_wide.csv'")

Data yang sudah dibersihkan (format wide) berhasil disimpan ke 'cleaned_financial_data_wide.csv'


In [ ]:
df_long.to_csv('cleaned_financial_data_long.csv', index=False, encoding='utf-8')
print("Data yang sudah dibersihkan (format long) berhasil disimpan ke 'cleaned_financial_data_long.csv'")

Data yang sudah dibersihkan (format long) berhasil disimpan ke 'cleaned_financial_data_long.csv'
